File: design-pattern-java/src/main/java/org/design/pattern/creacionales/pattern01/Pattern01Builder.java
package org.design.pattern.creacionales.pattern01;

import lombok.Getter;
import lombok.Setter;
import org.design.pattern.util.ConsoleUtil;
import org.fusesource.jansi.Ansi.Color;

/**
 * ! Patrón Builder: Es un patrón de diseño creacional que nos permite construir objetos complejos
 * paso a paso.
 * <p>
 * El patrón nos permite producir distintos tipos y representaciones de un objeto empleando el mismo
 * código de construcción.
 * <p>
 * Es útil cuando necesitamos construir un objeto complejo con muchas partes y queremos que el
 * proceso de construcción sea independiente de las partes que lo componen.
 */

public class Pattern01Builder {

    public static void main(String[] args) {
        ConsoleUtil.initialize();

        Computer basicComputer = new ComputerBuilder().setCpu("Intel core 2 duo")
            .setRam("4GB")
            .setStorage("256GB")
            .build();

        ConsoleUtil.println("Computadora básica:", Color.BLUE);
        basicComputer.displayConfiguration();

        Computer gamerComputer = new ComputerBuilder().setCpu("i7 14700k")
            .setCpu("i9 14900k")
            .setRam("128GB")
            .setStorage("4TB")
            .setGpu("RTX 5090 Ti")
            .build();

        ConsoleUtil.println("Computer gamer:", Color.RED);
        gamerComputer.displayConfiguration();

        ConsoleUtil.clean();
    }
}

@Getter
@Setter
class Computer {

    private String cpu;
    private String ram;
    private String storage;
    private String gpu;

    public void displayConfiguration() {
        String gpu = this.gpu != null ? this.gpu : "Sin GPU";

        ConsoleUtil.println(String.format(
            """
                Configuracion de la computadora:
                CPU: %s
                RAM: %s
                Almacenamiento: %s
                GPU: %s
                """, cpu, ram, storage, gpu
        ));
    }
}

class ComputerBuilder {

    private final Computer computer;

    public ComputerBuilder() {
        this.computer = new Computer();
    }

    public ComputerBuilder setCpu(String cpu) {
        computer.setCpu(cpu);
        return this;
    }

    public ComputerBuilder setRam(String ram) {
        computer.setRam(ram);
        return this;
    }

    public ComputerBuilder setStorage(String storage) {
        computer.setStorage(storage);
        return this;
    }

    public ComputerBuilder setGpu(String gpu) {
        computer.setGpu(gpu);
        return this;
    }

    public Computer build() {
        return computer;
    }
}

File: design-pattern-kotlin/src/main/kotlin/creacionales01/Pattern01Builder.kt
package creacionales01

class Computer(
    var cpu: String = "cpu - not defined",
    var ram: String = "ram - not defined",
    var storage: String = "storage - not defined",
    var gpu: String? = null
) {
    fun displayConfiguration() {
        print(
            """Configuración de la computadora: 
            CPU: ${this.cpu}
            RAM: ${this.ram}
            Almacenamiento: ${this.storage}
            GPU: ${if (this.gpu != null) this.gpu else "Sin gpu"}
        """
        )
    }
}

class ComputerBuilder {
    val computer: Computer = Computer()

    fun setCpu(cpu: String): ComputerBuilder {
        computer.cpu = cpu
        return this
    }

    fun setRam(ram: String): ComputerBuilder {
        computer.ram = ram
        return this
    }

    fun setStorage(storage: String): ComputerBuilder {
        computer.storage = storage
        return this
    }

    fun setGpu(gpu: String): ComputerBuilder {
        computer.gpu = gpu
        return this
    }

    fun build(): Computer {
        return computer
    }
}

fun main() {
    val basicComputer: Computer =
        ComputerBuilder().setCpu("Intel Core 2 Duo").setRam("4GB").setStorage("256GB").build()

    println("Computadora básica: ")
    basicComputer.displayConfiguration()

    val gamerComputer: Computer =
        ComputerBuilder().setCpu("i7 14700k").setCpu("i9 14900k").setRam("128GB").setStorage("4TB")
            .setGpu("RTX 5090 Ti").build()

    println("Computadora gamer:")
    gamerComputer.displayConfiguration()
}